frontend/packages/operator-lifecycle-manager/src/components/create-operand.tsx

```diff
@@ -104,297 +178,465 @@ const defaultValueFor = (field: OperandField) => {
-
-  const [k8sObj, setK8sObj] = React.useState<K8sResourceKind>(sample);
-  React.useEffect(() => {
-    setK8sObj((current) => {
-      const specValues = fields.reduce((usedFormValues, field) => {
-        const formValue = _.get(usedFormValues, field.path);
-        if (_.isEqual(formValue, defaultValueFor(field)) || _.isNil(formValue)) {
-          return _.omit(usedFormValues, field.path);
-        }
-        return usedFormValues;
-      }, _.omit(formValues, ['metadata.name', 'metadata.labels']));
+  const updateFormValues = React.useCallback((path: string, value: any) => {
+    setFormValues((current) => {
       const next = {
-        apiVersion: apiVersionForModel(operandModel),
-        kind: operandModel.kind,
-        metadata: {
-          namespace,
-          name: formValues['metadata.name'],
-          labels: SelectorInput.objectify(
-            formValues['metadata.labels'],
-          ) as ObjectMetadata['labels'],
-          annotations: _.get(sample, 'metadata.annotations', {}),
-        },
-        spec: _.reduce(
-          specValues,
-          (spec, value, path) => _.set(spec, path, value),
-          _.get(sample, 'spec', {}),
-        ),
+        ...current,
+        [path]: value,
       };
       if (_.isEqual(current, next)) {
         return current;
       }
-      onChange(next);
       return next;
     });
-  }, [fields, formValues, namespace, onChange, operandModel, sample]);
+  }, []);
 
-  const updateFormValues = (path: _.PropertyPath, value: any) => {
-    setFormValues((current) => {
-      const next = _.set(_.cloneDeep(current), path, value);
-      return _.isEqual(current, next) ? current : next;
-    });
-  };
+  // Build K8s object from current state
+  const formData = React.useMemo(() => {
+    // Build a k8s object from the current form fields and values in state
+    const newSpec = _.reduce(
+      fields,
+      (specAccumulator, field) => {
+        const value = _.get(formValues, field.path);
+        if (_.isNil(value) || _.isEqual(value, defaultValueFor(field.capabilities))) {
+          return specAccumulator;
+        }
+        return _.set(specAccumulator, field.path, value);
+      },
+      {},
+    );
 
+    return {
+      apiVersion: apiVersionForModel(operandModel),
+      kind: operandModel.kind,
+      metadata: {
+        namespace,
+        name: formValues['metadata.name'],
+        labels: SelectorInput.objectify(formValues['metadata.labels']) as ObjectMetadata['labels'],
+        annotations: _.get(buffer, 'metadata.annotations', {}),
+      },
+      spec: _.mergeWith(newSpec, buffer.spec, (newValue) => {
+        // Always replace old array with new one because if old array.length > new array.length,
+        // you will end  up with extra array field groups.
+        if (_.isArray(newValue)) {
+          return newValue;
+        }
+      }),
+    };
+  }, [fields, formValues, namespace, operandModel, buffer]);
+
+  const onSwitchToYAML = React.useCallback(() => {
```

Comment:
This is completely new. I moved the button for switching between form and yaml into the form and yaml components so that I could bubble up the current form or yaml data to the parent component only when the button is clicked. Before, the data was being updated on every change, which made syncing difficult and decreased performance.